Add support for custom blocks in .vue files

asselin opened the pull request:
This builds on https://github.com/vuejs/vue/pull/4157 adding support in vue-loader for custom blocks.  If a custom block is found in a .vue file, the loader for it will be looked up in the webpack vue configuration block just like a language is for script/style/template.

I think this improvement could probably cover the requests in some other feature requests, such as #457, #423, #413, #223, and having the ability in a UI kit project to specify (e.g.) Sass in a .vue file for a GUI component that is meant to be extracted and concatenated, and compiled into CSS by the application that pulls in and uses the UI kit.

This is a pretty simplistic implementation-- if there are issues that you think should be addressed in order to merge this, please let me know.

yyx990803 commented on lib/loader.js:
```diff
@@ -271,6 +273,20 @@ module.exports = function (content) {
   }
 
+  // add requires for customBlocks
+  if (parts.customBlocks && parts.customBlocks.length) {
```
We should avoid adding any code to the file unless there is a matching loader for the custom block.

yyx990803 commented on lib/loader.js:
```diff
@@ -155,6 +155,8 @@ module.exports = function (content) {
         case 'script':
           return injectString + ensureBang(ensureLoader(lang))
+        default:
+          return ensureBang(loaders[type])
```
This would probably throw if `loaders[type]` doesn't exist

kazupon replied:
Did you code updated?

asselin replied:
Sorry, should have replied to this.  This won't throw because there's a guard on line 289 that emits a warning in this case.  This is covered by the `warning should be raised when a loader is not specified for a custom block` unit test.

yyx990803 commented:
Thanks for the PR! A few issues that need to be fixed, and ideally we need a test case for this too.

asselin commented:
Thanks for the feedback!  I fixed the 2 issues and added them to the PR.  I'll work on adding some tests and updating the documentation one night this week.

asselin commented:
Hmmm.  I think I messed up my rebase... I'll try fixing that so the patch is cleaner...

asselin commented:
FYI, I haven't forgotten about building unit tests for this PR; I've just been slammed with a number of things demanding my time right now.  I still intend to finish this up, targeting the week of Dec 26-30.

asselin commented:
Unit tests added, please let me know if there are any other changes required.

iamkether commented:
@asselin this is looking great. I have one question though, would there be a way to 'ignore' certain custom blocks in different Webpack builds? Perhaps this is just me missing something in the documentation, or maybe it's something that needs to be added but, as an example, I wouldn't necessarily want to have vue-loader parsing documentation or unit-tests when I'm in the middle of development. 

I guess maybe with Webpack's caching system this might not make a huge performance issue, but even when thinking about CI systems, it would be nice to be able to have some way to "ignore" certain custom blocks.

asselin commented:
@iamkether If you want to leave some part of the .vue file out of the generated code, you can specify to use the null-loader in the webpack config file (for example, omit unit tests from debug & release builds).  vue-loader will still process that section of the .vue file, but passing it to the null loader will effectively make it cease processing there.

asselin commented:
@yyx990803 or @kazupon: This PR is ready for a 2nd review (and merge, hopefully).

dokether commented:
@asselin ah, okay, that makes sense, thank you.

iamkether commented:
@asselin ah, okay, that makes sense, thank you.

kazupon commented on docs/en/start/spec.md:
```diff
@@ -22,6 +22,10 @@ export default {
 </style>
+
+<custom1>
+	This could be e.g. documentation for the component.
```
Do we need indent?

asselin replied:
It was indented by a tab.  Changed to 2 spaces.

kazupon commented on docs/en/start/spec.md:
```diff
@@ -66,6 +70,66 @@ More details can be found in [Using Pre-Processors](../configurations/pre-proces
 
+### Custom Blocks
+
+Additional custom blocks can be included in a `*.vue` file for any project specific needs.  `vue-loader` will use the tag name to look up which webpack loaders should be applied to the contents of the section.  The webpack loaders should be specified in the `loaders` hash of the `vue` section of the webpack configuration in the same way that languages are specified for the standard sections of th ...
```
Please arrange the space.

asselin replied:
Sorry, I'm not clear on what you're asking for here.  Is it 1 space between sentences instead of 2?  If so, I just changed that and will push.

kazupon replied:
sorry, my poor explanation. :bow:
exactly!

asselin commented:
@kazupon I made updates to the documentation, and see my comment about the test not being necessary.  If I didn't make the changes you were looking for on the doc, please let me know.  Thanks!

kazupon commented:
@asselin Thanks! LGTM!

/ping @yyx990803
Please check just in case.

asselin commented:
@yyx990803 Is this good to merge now?  Any other cleanup needed?

FYI, I have a follow-on PR that I'm going to submit that adds support for adding attribute values from custom blocks to the webpack command as query parms.

Thanks!

yyx990803 commented:
@asselin sorry for dragging on, just busy with other stuff and don't want to rush-review this. Will try to finalize this week or next.

asselin commented:
Rebased in on current master, and added a fix for extract-text-plugin 2.0 rc

asselin commented:
@yyx990803 I see that you're spending some time on vue-loader; any chance this PR could be merged soon?
